plugins/user.py

```diff
@@ -0,0 +1,220 @@
+from will.plugin import WillPlugin
+from will.decorators import respond_to
+from will import settings
+import pyotp
+import qrcode
+import boto
+import boto.s3.connection
+from boto.s3.key import Key
+import pprint
+import cStringIO
+
+def check_user_permission(storage, username, permission):
+    user_permissions = storage.load("user_permissions", {})
+    if len(user_permissions) == 0:
+        if hasattr(settings, "ADMIN_USERS"):
+            for admin_user in settings.ADMIN_USERS.split(','):
+                user_permissions[admin_user] = ['admin', 'grant']
+                storage.say("giving {} admin permissions".format(admin_user))
+            storage.save("user_permissions", user_permissions)
+        else:
+            msg = "No admin users are defined in redis or environment"
+            storage.say(msg)
+    if username in user_permissions:
+        if permission in user_permissions[username]:
+            return 1
+        else:
+            return 0
+    else:
+        return 0
+
+def verify_twofactor(storage, username, token):
+    user_twofactor = storage.load("user_twofactor", {})
+    if username in user_twofactor:
+        totp = pyotp.TOTP(user_twofactor[username])
+        if totp.verify(token):
+            return 1
+        else:
+            return 0
+    else:
+        return 0
+
+class UserPlugin(WillPlugin):
+    def __init__(self):
+        if not hasattr(settings, "TWOFACTOR_ISSUER"):
+            msg = "Error: TWOFACTOR_ISSUER not defined in the environment"
+            self.say(msg)
+        self.TWOFACTOR_ISSUER = settings.TWOFACTOR_ISSUER
+        if not hasattr(settings, "TWOFACTOR_PRINCIPLE"):
+            msg = "Error: TWOFACTOR_PRINCIPLE not defined in the environment"
+            self.say(msg)
+        self.TWOFACTOR_PRINCIPLE = settings.TWOFACTOR_PRINCIPLE
+        if not hasattr(settings, "TWOFACTOR_S3_BUCKET"):
+            msg = "Error: TWOFACTOR_S3_BUCKET not defined in the environment"
+            self.say(msg)
+        self.TWOFACTOR_S3_BUCKET = settings.TWOFACTOR_S3_BUCKET
+        if not hasattr(settings, "TWOFACTOR_S3_PROFILE"):
+            msg = "Error: TWOFACTOR_S3_PROFILE not defined in the environment"
+            self._say_error(msg)
+        self.TWOFACTOR_S3_PROFILE = settings.TWOFACTOR_S3_PROFILE
+
+
+            
+
+
+    def generate_and_upload_QR(self,secret,username):
+        try:
+            principle = self.TWOFACTOR_PRINCIPLE
```

Comment:
You should probably kill the plugin on `__init__` if these things aren't defined, instead of checking them again later. Alternatively, just throw all of the checking in a function somewhere else.